File: leetcode/range-sum-query-2d-immutable.py
# https://leetcode.com/problems/range-sum-query-2d-immutable/

# Tags: Array - Design - Matrix - Prefix Sum

import timeit
from typing import List


# The naive approach would be to assign the matrix and calculate the result
# on each call to the sumRegion.
#
# Time complexity: O(n^2)
# Space complexity: O1
#
# Time limit exceeded on LeetCode.
class Naive:
    def __init__(self, matrix: List[List[int]]):
        self.matrix = matrix

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        sum = 0
        for row in range(row1, row2 + 1):
            for col in range(col1, col2 + 1):
                sum += self.matrix[row][col]
        return sum


# Create a matrix of prefix sums when the class is instantiated.
# Then use the prefix sums to calculate the sum of any submatrix in O(1)
#
# Time complexity of __init__: O(n) - we visit each element of the matrix to create the prefix sum.
# Time complexity of sumRegion: O(1) - using the prefix sum we can access in constant time a maximum of 4 matrix
# elements to create the sum of all elements in the submatrix.
# Space complexity: O(n) - We keep a matrix of size n in memory
#
# Runtime: 1808 ms, faster than 82.98% of Python3 online submissions for Range Sum Query 2D - Immutable.
# Memory Usage: 24.6 MB, less than 88.71% of Python3 online submissions for Range Sum Query 2D - Immutable.
class PrefixSum:
    def __init__(self, matrix: List[List[int]]):
        # Exit if no matrix passed in as input
        if matrix == None or not matrix or matrix[0] == None or not matrix[0]:
            return
        self.sums = [[0] * len(matrix[0]) for _ in range(len(matrix))]
        for row_idx, row in enumerate(matrix):
            row_sum = 0
            for col_idx, col in enumerate(row):
                row_sum += col
                self.sums[row_idx][col_idx] = row_sum if row_idx == 0 else row_sum + self.sums[row_idx - 1][col_idx]

        print(self.sums)

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        # Start with the total at the bottom-right edge of the selected area
        result = self.sums[row2][col2]
        # If the start row is not the first one, subtract the sum at (row1-1,col2)
        if row1 > 0:
            result -= self.sums[row1 - 1][col2]
        # If the start col is not the first one, subtract the sum at (row2, col1-1)
        if col1 > 0:
            result -= self.sums[row2][col1 - 1]
        # If both the start column and the start row are not the first one, we
        # have subtracted that area twice, we have to re-add its value
        if row1 > 0 and col1 > 0:
            result += self.sums[row1 - 1][col1 - 1]
        return result


def test():
    executors = [Naive, PrefixSum]
    tests = [
        [
            [
                [3, 0, 1, 4, 2],
                [5, 6, 3, 2, 1],
                [1, 2, 0, 1, 5],
                [4, 1, 0, 1, 7],
                [1, 0, 3, 0, 5],
            ],
            {  # Sum: submatrix vertices
                8: [2, 1, 4, 3],
                11: [1, 1, 2, 2],
                12: [1, 2, 2, 4],
            },
        ]
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1"))):
            for i, t in enumerate(tests):
                sol = executor(t[0])
                for sum in t[1]:
                    result = sol.sumRegion(*t[1][sum])
                    exp = sum
                    assert (
                        result == exp
                    ), f"\033[93m» {result} <> {exp}\033[91m for test {i} using \033[1m{executor.__name__}"
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        res = "{0:20}{1:10}{2:10}".format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/range-sum-query-2d-immutable.rs
// 304. Range Sum Query 2D - Immutable
// 🟠 Medium
//
// https://leetcode.com/problems/range-sum-query-2d-immutable/
//
// Tags: Array - Design - Matrix - Prefix Sum

struct NumMatrix {
    sums: Vec<Vec<i32>>,
}

/// Compute a matrix of prefix sums where sums[i][j] is the sum of all elements
/// between sums[0][0] and sums[i][j] including. Store this prefix sums matrix
/// in the struct. When we receive a query [r1, c1, r2, c2] we can compute in
/// O(1) as the sum of all elements up to r2, c2 minus the columns to the left,
/// if any, of c1, minus the rows above, if any of r1, if there is an area
/// above and to the left of r1, c1, we will have subtracted it twice, we add
/// it once and we have the range sum.
///
/// Time complexity: O(m*n) - For the new method, we read all elements of the
/// input and create a matrix of the same size with prefix sums. For the
/// sum_region method, the time complexity is O(1).
/// Space complexity: O(m*n) - We store a matrix of prefix sums of the same
/// size as the input matrix.
///
/// Runtime 102 ms Beats 100%
/// Memory 13.7 MB Beats 65.52%
impl NumMatrix {
    fn new(matrix: Vec<Vec<i32>>) -> Self {
        let mut sums: Vec<Vec<i32>> = vec![];
        for i in 0..matrix.len() {
            let mut row = vec![];
            let mut row_sum = 0;
            for j in 0..matrix[0].len() {
                row_sum += matrix[i][j];
                row.push(row_sum);
                if i > 0 {
                    row[j] += sums[i - 1][j]
                }
            }
            sums.push(row);
        }
        Self { sums }
    }

    fn sum_region(&self, row1: i32, col1: i32, row2: i32, col2: i32) -> i32 {
        let (row1, col1) = (row1 as usize, col1 as usize);
        let (row2, col2) = (row2 as usize, col2 as usize);
        // Start with the sum at the bottom-right.
        let mut range_sum = self.sums[row2][col2];
        // If the range does not go all the way left, remove these columns.
        if col1 > 0 {
            range_sum -= self.sums[row2][col1 - 1];
        }
        // If the range does not go all the way to the top, remove these rows.
        if row1 > 0 {
            range_sum -= self.sums[row1 - 1][col2];
        }
        // If we removed twice the top-left cells, add them once.
        if col1 > 0 && row1 > 0 {
            range_sum += self.sums[row1 - 1][col1 - 1]
        }
        range_sum
    }
}

// Tests.
fn main() {
    let num_matrix = NumMatrix::new(vec![
        vec![3, 0, 1, 4, 2],
        vec![5, 6, 3, 2, 1],
        vec![1, 2, 0, 1, 5],
        vec![4, 1, 0, 1, 7],
        vec![1, 0, 3, 0, 5],
    ]);
    assert_eq!(num_matrix.sum_region(2, 1, 4, 3), 8);
    assert_eq!(num_matrix.sum_region(1, 1, 2, 2), 11);
    assert_eq!(num_matrix.sum_region(1, 2, 2, 4), 12);
    println!("All tests passed!")
}
